ecommerce/models.py

```diff
@@ -154,3 +160,93 @@ def save(self, *args, **kwargs):
     def __str__(self):
         """Description for CoursePrice"""
         return "CoursePrice for {}, price={}, is_valid={}".format(self.course_run, self.price, self.is_valid)
+
+
+class Coupon(Model):
+    """
+    Model for a coupon. This stores the discount for the coupon, how many people can use it, and how many times
+    a person can use it.
+
+    When a coupon is redeemed by a purchaser the counter on this object is decremented
+    and a UserCoupon object is created for that particular purchaser.
+    """
+    PERCENT_DISCOUNT = 'percent-discount'
+    FIXED_DISCOUNT = 'fixed-discount'
+
+    AMOUNT_TYPES = [PERCENT_DISCOUNT, FIXED_DISCOUNT]
+
+    # The coupon code used for redemption by the purchaser in the user interface.
+    # If blank, the purchaser may not redeem this coupon through the user interface,
+    # though it may be redeemed in their name by an administrator.
+    coupon_code = TextField(null=True, blank=True)
+    # One and only one of these three foreign keys must be set to not null
+    course_run = ForeignKey(CourseRun, on_delete=SET_NULL, null=True)
+    course = ForeignKey(Course, on_delete=SET_NULL, null=True)
+    program = ForeignKey(Program, on_delete=SET_NULL, null=True)
+
+    # percent or fixed discount
+    amount_type = CharField(
+        choices=[(_type, _type) for _type in AMOUNT_TYPES],
+        max_length=30,
+    )
+    # Either a number from 0 to 1 representing a percent, or a fixed value for discount
+    amount = DecimalField(decimal_places=2, max_digits=20)
+
+    # Number of people this coupon can be redeemed by
+    num_coupons_available = IntegerField(null=False)
+    # Number of times a person can redeem a coupon
+    num_redemptions = IntegerField(null=False)
+    # After this time the coupons will not be redeemable
+    expiration_date = DateTimeField(null=True)
+    # If true, coupons are not presently redeemable
+    disabled = BooleanField(default=False)
+
+    def __str__(self):
+        """Description for Coupon"""
+        product = self.course_run or self.course or self.program
+        return "Coupon {amount_type} {amount} for {product}, {num_coupons_available} left".format(
+            amount_type=self.amount_type,
+            amount=self.amount,
+            product=product,
+            num_coupons_available=self.num_coupons_available,
+        )
+
+
+class UserCoupon(Model):
+    """
+    Model for uses of a coupon by a user.
+    """
+    coupon = ForeignKey(Coupon, on_delete=SET_NULL, null=True)
```

Comment:
A `UserCoupon` is a coupon which has already been redeemed by a user one or more times. Admittedly the name could be better, but I don't think an inheritance relationship fits here since the `UserCoupon` only links a `Coupon` to a `User` and stores how many times the user has used it.